docs/design-documents/platform/pin-names/pin_names.md

```diff
@@ -0,0 +1,198 @@
+# Pin Names design document
+
+# Table of contents
+
+1. [Table of contents](#table-of-contents).
+    * [Revision history](#revision-history).
+1. [Introduction](#introduction).
+    1. [Overview and background](#overview-and-background).
+    1. [Requirements and assumptions](#requirements-and-assumptions).
+1. [Detailed design](#detailed-design).
+    1. [Arduino Uno (Rev 3) Header Pins](#arduino-uno-rev-3-header-pins).
+        1. [I2C Definition for Arduino Uno (Rev3) Pins](#i2c-definition-for-arduino-uno-rev3-pins)
+        1. [SPI Definition for Arduino Uno (Rev3) Pins](#spi-definition-for-arduino-uno-rev3-pins)
+    1. [Custom/Vendor Header Pins](#custom/vendor-header-pins).
+        1. [Custom/Vendor Connector I2C Definition](#custom/vendor-connector-i2c-definition).
+        1. [Custom/Vendor Connector SPI Definition](#custom/vendor-connector-spi-definition).
+    1. [Board Components Pins](#board-components-pins).
+        1. [LED Definition](#led-definition)
+        1. [BUTTON Definition](#button-definition)
+1. [Other information](#other-information).
+    1. [Avoid Unnecessary Definitions](#avoid-unnecessary-definitions).
+
+
+### Revision history
+
+1.0 - Initial revision - Malavika Sajikumar - 3/5/2020.  
+This document was written for MBED OS version 5.15 (released on 11th of February 2020).
+
+# Introduction
+
+### Overview and background
+
+Mbed is designed so that application code written in the platform is portable across different Mbed supported boards with the same hardware capabilities or interfaces. However, the code, in most cases, is not truly portable due to the differences in pin name definitions for the same kind of interfaces across different boards. 
+
+This document provides a standard for defining pin names for the commonly used Mbed interfaces like I2C, SPI, etc. and board components like buttons and LEDs in a board support package.
+
+### Requirements and assumptions
+
+This document mainly applies to the pin standards required for Arduino Uno header which is shared among multiple Mbed enabled boards. The Arduino Uno header has been stable since 2012 at its current revision, which is the Arduino Uno Rev3. All design choices in this document for the Arduino Uno header are based on the Arduino Uno Rev3 header implementation.
+
+# Detailed design
+
+To achieve meaningful portability of application code across various Mbed enabled boards, the pin names of commonly used interfaces and board components should be common across these boards. This document describes a set of rules on how to name these pins in the board support package, specifically in PinNames.h file of the board.
+
+### Arduino Uno (Rev3) Header Pins
+
+All Arduino Uno (Rev3) form factor Mbed controller boards should define D0-D15 & A0-A5 pins as a default standard.
```

Comment:
They also need to make sure that `"ARDUINO"` is in the `supported-form-factors` section in the target.  That should at least get a mention.

That causes `TARGET__FF_ARDUINO` to be defined, and then lots of existing in-tree code will assume the D0-D15 and A0-A6 pins are present. (Eg drivers that have a default pin selection for Arduino form factor).

Reply:
(I should say that Malavika, who created this pull request, is a teammate of mine and I'm working with her on this proposal)

Agree.
One further point is it might be an idea to change the form factor name to ARDUINO_UNO and perhaps then TARGET_FF_ARDUINO to TARGET_FF_ARDUINO_UNO. There are Mbed supported Arduino Nano form-factor boards and maybe MKR form-factor as well (if there isn't already then the "Portenta H7" board will change that).

Reply:
> They also need to make sure that `"ARDUINO"` is in the `supported-form-factors` section in the target. That should at least get a mention.
> 
> That causes `TARGET__FF_ARDUINO` to be defined, and then lots of existing in-tree code will assume the D0-D15 and A0-A6 pins are present. (Eg drivers that have a default pin selection for Arduino form factor).

https://github.com/ARMmbed/mbed-os/pull/12838/commits/63a3c8f59c24f96e2d2f2bf9dec17f4671f86571

Reply:
> They also need to make sure that `"ARDUINO"` is in the `supported-form-factors` section in the target. That should at least get a mention.
> 
> That causes `TARGET__FF_ARDUINO` to be defined, and then lots of existing in-tree code will assume the D0-D15 and A0-A6 pins are present. (Eg drivers that have a default pin selection for Arduino form factor).

@kjbracey-arm Is it A0-A6 or A0-A5?

Reply:
I'd suggest that we introduce a new `"ARDUINOUNO"` form factor that follows the proposal being discussed here and we aim to deprecate the existing `"ARDUINO"` when possible. This should help us to ensure that new boards that follow the standard can be tested accordingly and we have a clear definition of pin-names. Let me know your thoughts on this.